### ISSUE TITLE
hs.window:setFrame() does not guard against absurd values

### ISSUE THREAD

**Issue by gibiansky** [Hammerspoon/hammerspoon#499]
Hammerspoon crashes and somehow causes the OS to crash on `setFrame`, when the arguments to `setFrame` are absurd. For example, something like this:

``` lua
local rect = {
  h = 1152000000.0,
  w = 2359296000.0,
  x = 4718592000.0,
  y = 33120000.0
}
hs.window.focusedWindow():setFrame(rect)
```

**Comment by gibiansky** [comment]
I actually can't get it to crash the OS quite as drastically with my test code, but instead on this code Hammerspoon takes up 100% CPU and 50% of my 16 GB of memory.

**Comment by cmsj** [comment]
@gibiansky Out of interest, how did you come across it? Did you just decide to try absurd values, or did some kind of bug end up with those values being chosen for you?

I can see the argument behind guarding ourselves against absurd values, but my question then is, how do you define absurd. The total size of all active displays? 2x that? 4x that?

**Comment by gibiansky** [comment]
Yes, it was a bug in my scaling code. My issue actually resulted in an OS crash somehow, I think caused by this. Was very strange.

Does a setFrame for a size larger than the display actually work? I thought it just clipped it to the maximum size. If that's right, then Hammerspoon should just do the clipping earlier.

**Comment by lowne** [comment]
It's not so straightforward... in fact, it would introduce quite a lot of overhead for every call, since HS itself doesn't keep state (enumerate screens, get frames, compute union, compute intersection).
For reference, this is how you could do it (in the near future, hopefully, as both geometry and moses aren't merged yet) script-side when `frame` comes from "dangerous" calculations upstream:

```lua
w:setFrame(hs.moses.reduce(hs.screen.allScreens(),function(s,v)return s:union(v:frame())end,hs.geometry('0,0 0x0')):intersect(frame))
```

**Comment by lowne** [comment]
Should be addressed by using `:setFrameInScreenBounds()`, merged in #563

**Comment by cmsj** [comment]
(I'm going to close this out because I don't think there is any further work to be done in Hammerspoon. Please re-open if you disagree!)
